### PR TITLE
feat: support yaml config file

### LINKED ISSUE #1008: feature request: Add support for fluent-bit yaml configuration file mode
### Is your feature request related to a problem? Please describe.

do we have a plan to add the operator support for the [yaml configuration file](https://docs.fluentbit.io/manual/administration/configuring-fluent-bit/yaml/configuration-file#config_processors) mode for fluent bit? it will be nice to resolve back pressure (possibly) if the load is extremely high.

### Describe the solution you'd like

Add support for fluent-bit yaml configuration file mode

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by cw-Guo**
### Which issue(s) this PR fixes:

Fixes #1008

### Does this PR introduced a user-facing change?

```release-note
1. fluent-bit supports yaml config mode
2. fluent-bit supports intput processors and output processors
```

### Additional documentation, usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by benjaminhuo** [comment]
@cw-Guo Thanks, that's great to have yaml config files

cc @wanjunlei @wenchajun @Gentleelephant @adiforluls

**Comment by cw-Guo** [comment]
@benjaminhuo This is ready for review. But the tests are limited.

The idea behind this implementation can be found in the origin [issue# 1008](https://github.com/fluent/fluent-operator/issues/1008)

To support processor, a `map[string]interface{}` is used to avoid defining the structure for the processors. Because the available processors in the official documents are very limited and are subjected to change.

We might add a new CRD for processors in the furture. (In practice, any filter can be a processor, without errors but this usage is not documented in the offical documents.)

**Comment by cw-Guo** [comment]
To run it, a new image for fluent-bit need to be built.

File: apis/fluentbit/v1alpha2/clusterfluentbitconfig_types.go
Comment (adiforluls) [inline apis/fluentbit/v1alpha2/clusterfluentbitconfig_types.go:53]: What would be the default configuration format?

**Comment by adiforluls** [comment]
I left a comment in the proposal about an edge case. And would like to see some test cases here if possible (unit tests at-least).

File: apis/fluentbit/v1alpha2/clusterfluentbitconfig_types.go
Comment (cw-Guo) [reply apis/fluentbit/v1alpha2/clusterfluentbitconfig_types.go:53]: the default will be the classic mode.

**Comment by cw-Guo** [comment]
> I left a comment in the proposal about an edge case. And would like to see some test cases here if possible (unit tests at-least).

i will add more tests. For custom plugin, i will also try to incooperate it but i am afraid it will be hard.

**Comment by adiforluls** [comment]
> > I left a comment in the proposal about an edge case. And would like to see some test cases here if possible (unit tests at-least).
> 
> i will add more tests. For custom plugin, i will also try to incooperate it but i am afraid it will be hard.

Thoughts:

- Old users who have custom plugin config will be on classic style by default, so no confusion here
- A new user comes along, adds a custom plugin config in classic style and selects yaml style in the fbc CR. IMO this should result in operator throwing an error, if the user selects yaml style config, then they should provide yaml style configs in custom plugin and vice versa.

This is just my opinion, I'd let other maintainers pitch in as well on the PR.

**Comment by cw-Guo** [comment]
@adiforluls I can probably add one more field in the custom plugin to take in the unsupported plugins, just like the current processor implementation. 
https://github.com/fluent/fluent-operator/pull/1208/files#diff-487bd681e6fe01d15ade2e5a8650963714441e9282d9d6898c8c3ce1ab09aa43R76

**Comment by cw-Guo** [comment]
@benjaminhuo @adiforluls @wanjunlei @wenchajun @Gentleelephant 
May I have your reviews?

**Comment by benjaminhuo** [comment]
> @benjaminhuo @adiforluls @wanjunlei @wenchajun @Gentleelephant May I have your reviews?

Sure, we'll review this. Thanks for the contribution.
cc @wanjunlei @wenchajun @Gentleelephant

File: apis/fluentbit/v1alpha2/output_types.go
Comment (adiforluls) [inline apis/fluentbit/v1alpha2/output_types.go]: How are we handling classic config now for custom plugin? I see you have removed it in places and replaced it with yaml config.

Comment (cw-Guo) [reply apis/fluentbit/v1alpha2/output_types.go]: It will function as before, since it is just removed in the new loadAsYaml function, the old load function is not changed.

Comment (adiforluls) [reply apis/fluentbit/v1alpha2/output_types.go]: Oh I see this is func to load as yaml

**Review by adiforluls — DISMISSED** [review]
Given it touches a lot of files like fluent-watcher binaries, I'd definitely want other reviewers to go through the change as well, but in terms of overall configuration and cases LGTM.

File: apis/fluentbit/v1alpha2/clusterfilter_types.go
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/clusterfilter_types.go:22]: Unsorted imports.

File: apis/fluentbit/v1alpha2/plugins/custom/custom_types.go
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/plugins/custom/custom_types.go:6]: Unsorted imports.

File: apis/fluentbit/v1alpha2/clusterfluentbitconfig_types.go
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/clusterfluentbitconfig_types.go:22]: Unsorted imports.

File: apis/fluentbit/v1alpha2/clusteroutput_types.go
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/clusteroutput_types.go:22]: Unsorted imports.

File: apis/fluentbit/v1alpha2/plugins/params/kvs.go
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/plugins/params/kvs.go:6]: Unsorted imports.

File: cmd/fluent-watcher/fluentbit/main.go
Comment (wanjunlei) [inline cmd/fluent-watcher/fluentbit/main.go:65]: The classic config file path => The yaml config file path?

**Review by wanjunlei — CHANGES_REQUESTED** [review]

**Comment by wanjunlei** [comment]
I have some questions and suggestions.
- This PR does not involve namespace CRD. Will this be implemented in the future?
- In the long run, we don't need to maintain two sets of configuration files. We will remove the support for classic configuration files in future versions. So when the fluentbit version is greater than 2.0.0, can we consider using the YAML configuration file as the default configuration file?
- It is not a good idea to splice yaml files. I suggest defining a structure to generate yaml config file.

Just some suggestions, not blocking this pr.

**Comment by cw-Guo** [comment]
@wanjunlei ❤️ Thanks for the review and suggestions. I have fixed the import orders.

> This PR does not involve namespace CRD. Will this be implemented in the future?

I am not too sure about how the namespace CRDs works. I didn't use it in my previous experience. But I did add support for:

1. [Filter](https://github.com/fluent/fluent-operator/blob/d0c76789f4ed424d4fd98557de349cc667576785/apis/fluentbit/v1alpha2/filter_types.go#L108),
2. [Output](https://github.com/fluent/fluent-operator/blob/d0c76789f4ed424d4fd98557de349cc667576785/apis/fluentbit/v1alpha2/output_types.go)

They are passed in to [fluentbitconfig_controller.go](https://github.com/fluent/fluent-operator/blob/d0c76789f4ed424d4fd98557de349cc667576785/controllers/fluentbitconfig_controller.go#L165). From my understanding, this will be enough but please correct me if i understand it wronly.

> In the long run, we don't need to maintain two sets of configuration files... So when the fluentbit version is greater than 2.0.0, can we consider using the YAML configuration file as the default configuration file?

> It is not a good idea to splice yaml files. I suggest defining a structure to generate yaml config file.

I agree that we don't want to maintain two sets of configuration files and I am also against splicing yaml files. But I also hasitate to make the (hard) decision for the users. _Avoiding breaking changes is the start point of my design_.  I am afraid only supporting yaml config file format would possibly make some users abandon the project.

If the community has decided to only support the yaml config file format, I am happy to see that a refractor for fluent-operator or a newer api version.

**Comment by wanjunlei** [comment]
> @wanjunlei ❤️ Thanks for the review and suggestions. I have fixed the import orders.
> 
> > This PR does not involve namespace CRD. Will this be implemented in the future?
> 
> I am not too sure about how the namespace CRDs work. I didn't use it in my previous experience. But I did add support for:
> 
> 1. [Filter](https://github.com/fluent/fluent-operator/blob/d0c76789f4ed424d4fd98557de349cc667576785/apis/fluentbit/v1alpha2/filter_types.go#L108),
> 2. [Output](https://github.com/fluent/fluent-operator/blob/d0c76789f4ed424d4fd98557de349cc667576785/apis/fluentbit/v1alpha2/output_types.go)
> 
> They are passed in to [fluentbitconfig_controller.go](https://github.com/fluent/fluent-operator/blob/d0c76789f4ed424d4fd98557de349cc667576785/controllers/fluentbitconfig_controller.go#L165). From my understanding, this will be enough but please correct me if i understand it wronly.
> 
> > In the long run, we don't need to maintain two sets of configuration files... So when the fluentbit version is greater than 2.0.0, can we consider using the YAML configuration file as the default configuration file?
> 
> > It is not a good idea to splice yaml files. I suggest defining a structure to generate yaml config file.
> 
> I agree that we don't want to maintain two sets of configuration files and I am also against splicing yaml files. But I also hasitate to make the (hard) decision for the users. _Avoiding breaking changes is the start point of my design_. I am afraid only supporting yaml config file format would possibly make some users abandon the project.
> 
> If the community has decided to only support the yaml config file format, I am happy to see that a refractor for fluent-operator or a newer api version.

OK, got it. The namespace CRD just needs to change the FIlter and Output.

**Review by wanjunlei — APPROVED** [review]

**Comment by benjaminhuo** [comment]
> I agree that we don't want to maintain two sets of configuration files and I am also against splicing yaml files. But I also hasitate to make the (hard) decision for the users. _Avoiding breaking changes is the start point of my design_. I am afraid only supporting yaml config file format would possibly make some users abandon the project.
> 
> If the community has decided to only support the yaml config file format

The old config file format might exist for some time for the yaml format to mature and catch up.
There are quite a lot of things to catch up.

**Review by benjaminhuo — APPROVED** [review]

**Comment by benjaminhuo** [comment]
Thanks @cw-Guo very much for such a big enhancement! I've invite you as fluent-operator maintainer.
@patrick-stephens @agup006 would you help to invite @cw-Guo to the fluent org?

**Pull request merged**
